Add review changes for Build MegaService of ChatQnA on Gaudi README.md and associated files. (#1595)
opea-project/GenAIExamples: ChatQnA/docker_compose/intel/hpu/gaudi/README.md

9a9f3fd4fd4c   opea/vllm-gaudi:latest                                                                          "python3 -m vllm.ent…"   2 minutes ago   Exited (1) 2 minutes ago                                                                                           vllm-gaudi-server
1ab9bbdf5182   redis/redis-stack:7.2.0-v9                                                                      "/entrypoint.sh"         2 minutes ago   Up 2 minutes                0.0.0.0:6379->6379/tcp, :::6379->6379/tcp, 0.0.0.0:8001->8001/tcp, :::8001->8001/tcp   redis-vector-db
9ee0789d819e   ghcr.io/huggingface/text-embeddings-inference:cpu-1.5                                           "text-embeddings-rou…"   2 minutes ago   Up 2 minutes                0.0.0.0:8090->80/tcp, [::]:8090->80/tcp                                                tei-embedding-gaudi-server
```

### Test the Pipeline

Once the ChatQnA services are running, test the pipeline using the following command:

```bash

> curl http://${host_ip}:8888/v1/chatqna \

Reviewer: Need to explain where this `host_ip` is coming from.

| chatqna-gaudi-ui-server      | opea/chatqna-ui:latest                                | No       | Provides the user interface for the ChatQnA application.                                           |
| chatqna-gaudi-nginx-server   | opea/nginx:latest                                     | No       | Acts as a reverse proxy, managing traffic between the UI and backend services.                     |
| jaeger                       | jaegertracing/all-in-one:latest                       | Yes      | Provides tracing and monitoring capabilities for distributed systems.                              |

Many of these services provide pipeline support required for all ChatQnA deployments, and are not specific to supporting the Intel® Gaudi® platform. Therefore, while the `redis-vector-db`, `dataprep-redis-service`, `retriever`, `chatqna-gaudi-backend-server`, `chatqna-gaudi-ui-server`, `chatqna-gaudi-nginx-server`, `jaeger` are configurable, they will not be covered by this example, which will focus on the configuration specifics of the services modified to support the Intel® Gaudi® platform.

### vllm-service & tgi-service

> In the configuration of the `vllm-service` and the `tgi-service`, two primary variables play a primary role in determining the service's performance and functionality: `LLM_MODEL_ID` and `NUM_CARDS`. Both can be set using the appropriate environment variables. The `LLM_MODEL_ID` parameter specifies the particular large language model (LLM) that the service will utilize, effectively determining the capabilities and characteristics of the language processing tasks it can perform. This model identifier ensures that the service is aligned with the specific requirements of the application, whether it involves text generation, comprehension, or other language-related tasks. The `NUM_CARDS` parameter dictates the number of Gaudi devices allocated to the service. A higher number of Gaudi devices can enhance parallel processing capabilities, reduce latency, and improve throughput.

Reviewer 2: ```suggestion
In the configuration of the `vllm-service` and the `tgi-service`, two variables play a primary role in determining the service's performance and functionality: `LLM_MODEL_ID` and `NUM_CARDS`. Both can be set using the appropriate environment variables. The `LLM_MODEL_ID` parameter specifies the particular large language model (LLM) that the service will utilize, effectively determining the capabilities and characteristics of the language processing tasks it can perform. This model identifier ensures that the service is aligned with the specific requirements of the application, whether it involves text generation, comprehension, or other language-related tasks. The `NUM_CARDS` parameter dictates the number of Gaudi devices allocated to the service. A higher number of Gaudi devices can enhance parallel processing capabilities, reduce latency, and improve throughput.
```

### tei-embedding-service & tei-reranking-service

The `ghcr.io/huggingface/text-embeddings-inference:cpu-1.5` image supporting `tei-embedding-service` and `tei-reranking-service` depends on the `EMBEDDING_MODEL_ID` or `RERANK_MODEL_ID` environment variables respectively to specify the embedding model and reranking model used for converting text into vector representations and rankings. This choice impacts the quality and relevance of the embeddings rerankings for various applications. Unlike the `vllm-service`, the `tei-embedding-service` and `tei-reranking-service` each typically acquires only one Gaudi device and does not use the `NUM_CARDS` parameter; embedding and reranking tasks generally do not require extensive parallel processing and one Gaudi per service is appropriate. The list of [supported embedding and reranking models](https://github.com/huggingface/tei-gaudi?tab=readme-ov-file#supported-models) can be found at the the [huggingface/tei-gaudi](https://github.com/huggingface/tei-gaudi?tab=readme-ov-file#supported-models) website.

> ### tgi-gaurdrails-service

Reviewer 2: ```suggestion
### tgi-guardrails-service
```

This example includes the following sections:

- [ChatQnA Quick Start Deployment](#chatqna-quick-start-deployment): Demonstrates how to quickly deploy a ChatQnA platform on a Intel Gaudi platform.
- [ChatQnA Docker Compose Files](#chatqna-docker-compose-files): Describes the example docker compose files and the example deployments.
- [ChatQnA Service Configuration](#chatqna-service-configuration): Describes the services and possible configuration changes.

**Note** This example requires access to a properly installed Intel Gaudi platform with a functional Docker service configured to use the habanalabs-container-runtime. Please consult the [Intel Gaudi software Installation Guide](https://docs.habana.ai/en/v1.20.0/Installation_Guide/Driver_Installation.html) for more information.

## ChatQnA Quick Start Deployment

This section describes how to quickly deploy and test the ChatQnA service manually on an Intel Gaudi platform. The basic steps are:

1. [Access the Code](#access-the-code)

> 2. [Generate a HuggingFace Access Token](#generate-a-huggingface-access-token)

Reviewer 2: One could re-use a HuggingFace token too ...
[Obtain a HuggingFace token] ..

3. [Configure the Deployment Environment](#configure-the-deployment-environment)
4. [Deploy the Services Using Docker Compose](#deploy-the-services-using-docker-compose)
5. [Check the Deployment Status](#check-the-deployment-status)
6. [Test the Pipeline](#test-the-pipeline)
7. [Cleanup the Deployment](#cleanup-the-deployment)

### Access the Code

Clone the GenAIExample repository and access the ChatQnA Intel Gaudi platform Docker Compose files and supporting scripts:

> ```

Reviewer 2: Please ask them to checkout a release version, 1.2, etc after they git clone .. versus  using the bleeding edge of development.

git clone https://github.com/opea-project/GenAIExamples.git
cd GenAIExamples/ChatQnA/docker_compose/intel/hpu/gaudi/
```

### Generate a HuggingFace Access Token

> Setup a [HuggingFace](https://huggingface.co/) account and generate a [user access token](https://huggingface.co/docs/transformers.js/en/guides/private#step-1-generating-a-user-access-token), if required. An access token is required for some services to access model.

Reviewer 2: ```suggestion
Some HuggingFace resources, such as some models, are only accessible if you have an access token. If you do not already have a HuggingFace access token, you can create one by first creating an account by following the steps provided at [HuggingFace](https://huggingface.co/) account and then generating a [user access token](https://huggingface.co/docs/transformers.js/en/guides/private#step-1-generating-a-user-access-token).
```

### Configure the Deployment Environment

To set up environment variables for deploying ChatQnA services, source the _setup_env.sh_ script in this directory:

```
source ./set_env.sh
```

> The _set_env.sh_ script will prompt for required and optional environment variables used to configure the ChatQnA services. If a value is not entered the script will enable a default value for the environment variable. It will also generate a _.env_ file defining the desired configuration. Consult the section on [ChatQnA Service configuration](#chatqna-service-configuration) for information how service specific configuration parameters affect deployments.

Reviewer 2: ```suggestion
The _set_env.sh_ script will prompt for required and optional environment variables used to configure the ChatQnA services. If a value is not entered, the script will use a default value for the same. It will also generate a _.env_ file defining the desired configuration. Consult the section on [ChatQnA Service configuration](#chatqna-service-configuration) for information on how service specific configuration parameters affect deployments.
```

 ✔ Container tei-reranking-gaudi-server    Started                                                                      0.7s
 ✔ Container vllm-gaudi-server             Started                                                                      0.7s
 ✔ Container tei-embedding-gaudi-server    Started                                                                      0.3s
 ✔ Container redis-vector-db               Started                                                                      0.6s
 ✔ Container retriever-redis-server        Started                                                                      1.1s
 ✔ Container dataprep-redis-server         Started                                                                      1.1s
 ✔ Container chatqna-gaudi-backend-server  Started                                                                      1.3s
 ✔ Container chatqna-gaudi-ui-server       Started                                                                      1.7s
 ✔ Container chatqna-gaudi-nginx-server    Started                                                                      1.9s
```

### Check the Deployment Status

> After running docker compose, check if all the containers launched via docker compose has started:

Reviewer 2: ```suggestion
After running docker compose, check if all the containers launched via docker compose have started:
```

```
docker ps -a
```

> For the default deployment, the following 10 containers should be started:

Reviewer 2: ```suggestion
For the default deployment, the following 10 containers should have started:
```

Reviewer 2: point out the release tag when it is used, how it appears or "latest".

 ✔ Container chatqna-gaudi-backend-server  Removed                                                                                                 10.3s
 ✔ Container vllm-gaudi-server             Removed                                                                                                  0.0s
 ✔ Container retriever-redis-server        Removed                                                                                                 10.4s
 ✔ Container tei-reranking-gaudi-server    Removed                                                                                                  2.0s
 ✔ Container tei-embedding-gaudi-server    Removed                                                                                                  1.2s
 ✔ Container redis-vector-db               Removed                                                                                                  0.4s
 ✔ Network gaudi_default                   Removed                                                                                                  0.4s
```

> The ChatQnA containers should be stopped and removed after execution finishes.

Reviewer 2: ```suggestion
All the ChatQnA containers will first stopped and then removed after the completion of the "down" command.

## ChatQnA Docker Compose Files

In the context of deploying a ChatQnA pipeline on an Intel Gaudi platform, the allocation and utilization of Gaudi devices across different services are important considerations for optimizing performance and resource efficiency. Each of the four example deployments, defined by the example Docker compose yaml files, demonstrates a unique approach to leveraging Gaudi hardware, reflecting different priorities and operational strategies.

### compose.yaml - Default Deployment

> The default deployment utilizes Gaudi devices primarily for the `vllm-service`, which handles large language model (LLM) tasks. This service is configured to maximize the use of Gaudi's capabilities, potentially allocating multiple devices to enhance parallel processing and throughput. The `tei-reranking-service` also uses Gaudi hardware (1 card), however, indicating a balanced approach where both LLM processing and reranking tasks benefit from Gaudi's performance enhancements.

Reviewer 2: may want them to note the name "gaudi" in the container name ghcr.io/huggingface/tei-gaudi:1.5.0 opea/vllm-gaudi:latest

| Service Name                 | Image Name                                            | Gaudi Use    |
| ---------------------------- | ----------------------------------------------------- | ------------ |
| redis-vector-db              | redis/redis-stack:7.2.0-v9                            | No           |
| dataprep-redis-service       | opea/dataprep:latest                                  | No           |
| tei-embedding-service        | ghcr.io/huggingface/text-embeddings-inference:cpu-1.5 | No           |
| retriever                    | opea/retriever:latest                                 | No           |
| tei-reranking-service        | ghcr.io/huggingface/tei-gaudi:1.5.0                   | 1 card       |
| vllm-service                 | opea/vllm-gaudi:latest                                | Configurable |
| chatqna-gaudi-backend-server | opea/chatqna:latest                                   | No           |
| chatqna-gaudi-ui-server      | opea/chatqna-ui:latest                                | No           |
| chatqna-gaudi-nginx-server   | opea/nginx:latest                                     | No           |

### compose_tgi.yaml - TGI Deployment

> The TGI (Text Generation Inference) deployment and the default deployment differ primarily in their service configurations and specific focus on handling large language models (LLMs). The TGI deployment includes a unique `tgi-service`, which utilizes the `ghcr.io/huggingface/tgi-gaudi:2.0.6` image and is specifically configured to run on Gaudi hardware. This service is designed to handle LLM tasks with optimizations such as `ENABLE_HPU_GRAPH` and `USE_FLASH_ATTENTION`. The `chatqna-gaudi-backend-server` in the TGI deployment depends on the `tgi-service`, whereas in the default deployment, it relies on the `vllm-service`.

Reviewer 2: avoid names with <2.0.6> instead say <tag>

| redis-vector-db              | redis/redis-stack:7.2.0-v9                            | No             |
| dataprep-redis-service       | opea/dataprep:latest                                  | No             |
| tei-embedding-service        | ghcr.io/huggingface/text-embeddings-inference:cpu-1.5 | No             |
| retriever                    | opea/retriever:latest                                 | No             |
| vllm-service                 | opea/vllm-gaudi:latest                                | Configurable   |
| chatqna-gaudi-backend-server | **opea/chatqna-without-rerank:latest**                | No             |
| chatqna-gaudi-ui-server      | opea/chatqna-ui:latest                                | No             |
| chatqna-gaudi-nginx-server   | opea/nginx:latest                                     | No             |

This setup might allow for more Gaudi devices to be dedicated to the `vllm-service`, enhancing LLM processing capabilities and accommodating larger models. However, it also means that the benefits of reranking are sacrificed, which could impact the overall quality of the pipeline's output.

> ### compose_guardrails.yaml - Gaurdrails Deployment

Reviewer 2: ```suggestion
### compose_guardrails.yaml - Guardrails Deployment
```

> ### compose_guardrails.yaml - Gaurdrails Deployment

Reviewer 3: "Gaurdrails" → "Guardrails" please fix the typo.

|                              | opea/chatqna-without-rerank:latest                    |          |                                                                                                    |
|                              | opea/chatqna-guardrails:latest                        |          |                                                                                                    |
| chatqna-gaudi-ui-server      | opea/chatqna-ui:latest                                | No       | Provides the user interface for the ChatQnA application.                                           |
| chatqna-gaudi-nginx-server   | opea/nginx:latest                                     | No       | Acts as a reverse proxy, managing traffic between the UI and backend services.                     |
| jaeger                       | jaegertracing/all-in-one:latest                       | Yes      | Provides tracing and monitoring capabilities for distributed systems.                              |

> Many of these services provide pipeline support required for all ChatQnA deployments, and are not specific to supporting the Intel Gaudi platform. Therefore, while the `redis-vector-db`, `dataprep-redis-service`, `retriever`, `chatqna-gaudi-backend-server`, `chatqna-gaudi-ui-server`, `chatqna-gaudi-nginx-server`, `jaeger` are configurable, they will not be covered by this example, which will focus on the configuration specifics of the services modified to support the Intel Gaudi platform.

Reviewer 3: Intel Gaudi -> Intel® Gaudi®

### vllm-service & tgi-service

In the configuration of the `vllm-service` and the `tgi-service`, two primary variables play a primary role in determining the service's performance and functionality: `LLM_MODEL_ID` and `NUM_CARDS`. Both can be set using the appropriate environment variables. The `LLM_MODEL_ID` parameter specifies the particular large language model (LLM) that the service will utilize, effectively determining the capabilities and characteristics of the language processing tasks it can perform. This model identifier ensures that the service is aligned with the specific requirements of the application, whether it involves text generation, comprehension, or other language-related tasks. The `NUM_CARDS` parameter dictates the number of Gaudi devices allocated to the service. A higher number of Gaudi devices can enhance parallel processing capabilities, reduce latency, and improve throughput.

> However, developers need to be aware of the models that have been tested with the appropriate service image supporting the `vllm-service` and `tgi-service`. For example, documentation for the OPEA GenAIComps v1.0 release specify the list of [validated LLM models](https://github.com/opea-project/GenAIComps/blob/v1.0/comps/llms/text-generation/README.md#validated-llm-models) for each Gaudi enabled service images. Specific models may have stringent requirements on the number of Intel Gaudi devices required to support them.

Reviewer 2: ```suggestion
However, developers need to be aware of the models that have been tested with the respective service image supporting the `vllm-service` and `tgi-service`. For example, documentation for the OPEA GenAIComps v1.0 release specify the list of [validated LLM models](https://github.com/opea-project/GenAIComps/blob/v1.0/comps/llms/text-generation/README.md#validated-llm-models) for each Gaudi enabled service image. Specific models may have stringent requirements on the number of Intel Gaudi devices required to support them.
```

#### Deepseek Model Support for Intel Gaudi Platform ChatQnA pipeline

> ChatQnA now supports running the latest DeepSeek models, including [deepseek-ai/DeepSeek-R1-Distill-Llama-70B](https://huggingface.co/deepseek-ai/DeepSeek-R1-Distill-Llama-70B) and [deepseek-ai/DeepSeek-R1-Distill-Qwen-32B](https://huggingface.co/deepseek-ai/DeepSeek-R1-Distill-Qwen-32B) on Gaudi accelerators. To run `deepseek-ai/DeepSeek-R1-Distill-Llama-70B`, set the `LLM_MODEL_ID` appropriately and the `NUM_CARDS` to 8. To run `deepseek-ai/DeepSeek-R1-Distill-Qwen-32B`, update the `LLM_MODEL_ID` appropriately and the `NUM_CARDS` to 4.

Reviewer 2: ```suggestion
ChatQnA now supports running the latest DeepSeek models, including [deepseek-ai/DeepSeek-R1-Distill-Llama-70B](https://huggingface.co/deepseek-ai/DeepSeek-R1-Distill-Llama-70B) and [deepseek-ai/DeepSeek-R1-Distill-Qwen-32B](https://huggingface.co/deepseek-ai/DeepSeek-R1-Distill-Qwen-32B) on Gaudi accelerators. To run `deepseek-ai/DeepSeek-R1-Distill-Llama-70B`, set the `LLM_MODEL_ID` appropriately and the `NUM_CARDS` to 8. To run `deepseek-ai/DeepSeek-R1-Distill-Qwen-32B`, update the `LLM_MODEL_ID` appropriately and set the `NUM_CARDS` to 4.
```
